applications/StructuralMechanicsApplication/custom_elements/truss_elements/linear_truss_element.cpp

```diff
@@ -186,21 +206,64 @@ void LinearTrussElement2D<TNNodes>::GetShapeFunctionsValuesY(
         rN.resize(SystemSize, false);
 
     rN.clear();
-    if constexpr (NNodes == 2) {
-        rN[1] = 0.5 * (1.0 - xi);
-        rN[3] = 0.5 * (1.0 + xi);
-    } else { // 3N
-        rN[1] = 0.5 * xi * (xi - 1.0);
-        rN[3] = 0.5 * xi * (xi + 1.0);
-        rN[5] = (1.0 - std::pow(xi, 2));
+    array_1d<double, NNodes> base_N;
+    noalias(base_N) = GetBaseShapeFunctions(xi);
+    
+    if constexpr (Dimension == 2) {
+        if constexpr (NNodes == 2) {
+            rN[1] = base_N[0];
+            rN[3] = base_N[1];
+        } else { // 3N
+            rN[1] = base_N[0];
+            rN[3] = base_N[1];
+            rN[5] = base_N[2];
+        }
+    } else {
+        if constexpr (NNodes == 2) {
+            rN[1] = base_N[0];
+            rN[4] = base_N[1];
+        } else { // 3N
+            rN[1] = base_N[0];
+            rN[4] = base_N[1];
+            rN[7] = base_N[2];
+        }
     }
 }
 
 /***********************************************************************************/
 /***********************************************************************************/
 
-template<SizeType TNNodes>
-void LinearTrussElement2D<TNNodes>::GetFirstDerivativesShapeFunctionsValues(
+template<SizeType TNNodes, SizeType TDimension>
+void LinearTrussElement<TNNodes, TDimension>::GetShapeFunctionsValuesZ(
+    SystemSizeBoundedArrayType& rN,
+    const double Length,
+    const double xi
+    ) const
+{
+    if (rN.size() != SystemSize)
+        rN.resize(SystemSize, false);
```

Comment:
The `.resize` checks for equal size itself before execution, I wonder why this check is repeated here and in many other places.

Reply:
Good question, according to @RiccardoRossi the `resize` method adds an overhead to the operation